test/common/reference_data_wrangler.hpp

```diff
@@ -269,28 +270,75 @@ void verify_dft(const portfft::descriptor<Scalar, Domain>& desc, std::vector<Ele
   for (std::size_t i = 0; i < dft_offset; ++i) {
     if (ref_output[i] != actual_output[i]) {
       if constexpr (!IsInterleaved) {
-        std::cerr << elem_name << " part:";
+        std::cerr << "real part:";
       }
       std::cerr << "Incorrectly written value in padding at global idx " << i << ", ref " << ref_output[i] << " vs "
                 << actual_output[i] << std::endl;
+
       throw std::runtime_error("Verification Failed");
     }
+    if constexpr (!IsInterleaved) {
+      if (ref_output_imag[i] != actual_output_imag[i]) {
+        std::cerr << "imag part:";
+        std::cerr << "Incorrectly written value in padding at global idx " << i << ", ref " << ref_output_imag[i]
+                  << " vs " << actual_output_imag[i] << std::endl;
+      }
+    }
+  }
+  portfft::detail::dump_host("ref_output:", ref_output.data(), ref_output.size());
+  portfft::detail::dump_host("actual_output:", actual_output.data(), actual_output.size());
+  if constexpr (!IsInterleaved) {
+    portfft::detail::dump_host("ref_output_imag:", ref_output_imag.data(), ref_output_imag.size());
+    portfft::detail::dump_host("actual_output_imag:", actual_output_imag.data(), actual_output_imag.size());
   }
 
+  Scalar max_L2_rel_err = 0;
   for (std::size_t t = 0; t < desc.number_of_transforms; ++t) {
     const ElemT* this_batch_ref = ref_output.data() + dft_len * t + dft_offset;
     const ElemT* this_batch_computed = actual_output.data() + dft_len * t + dft_offset;
+    const ElemT* this_batch_ref_imag = ref_output_imag.data() + dft_len * t + dft_offset;
+    const ElemT* this_batch_computed_imag = actual_output_imag.data() + dft_len * t + dft_offset;
 
+    Scalar L2_err = 0;
+    Scalar L2_norm = 0;
     for (std::size_t e = 0; e != dft_len; ++e) {
-      const auto diff = std::abs(this_batch_computed[e] - this_batch_ref[e]);
+      BwdType computed_val = this_batch_computed[e];
+      BwdType ref_val = this_batch_ref[e];
+      if constexpr (!IsInterleaved) {
+        computed_val += std::complex<Scalar>(0, this_batch_computed_imag[e]);
+        ref_val += std::complex<Scalar>(0, this_batch_ref_imag[e]);
+      }
+      Scalar err = std::abs(computed_val - ref_val);
+      Scalar norm_val = std::abs(ref_val);
+      L2_err += err * err;
+      L2_norm += norm_val * norm_val;
+    }
+    L2_err = std::sqrt(L2_err);
+    L2_norm = std::sqrt(L2_norm);
+    Scalar L2_rel_err = L2_err / L2_norm;
+    max_L2_rel_err = std::max(max_L2_rel_err, L2_rel_err);
+  }
+  std::cout << "Max (across batches) relative L2 error: " << max_L2_rel_err << std::endl;
```

Comment:
I see we have a new for loop to compute `max_L2_rel_err` and print it. How are we expected to use this?
I think it can be confusing to print a message related to an error even when a test is passing. Maybe that's something we could enable only if some logging is enabled?

Reply:
This is the value that is actually comparable to literature about other FFT libraries. I would prefer to keep this as it is until I start with the logging work.